pkg/integrations/v2/app_o11y_receiver/sourcemaps/sourcemaps.go

```diff
@@ -0,0 +1,331 @@
+		return store.downloadSourceMapContent(sourceURL)
+	}
+	return nil, "", nil
+}
+
+func (store *RealSourceMapStore) getSourceMap(sourceURL string, release string) (*sourceMap, error) {
+	store.Lock()
+	defer store.Unlock()
+
+	cacheKey := fmt.Sprintf("%s__%s", sourceURL, release)
+
+	if smap, ok := store.cache[cacheKey]; ok {
+		return smap, nil
+	}
+	content, sourceMapURL, err := store.getSourceMapContent(sourceURL, release)
+	if err != nil || content == nil {
+		store.cache[cacheKey] = nil
+		return nil, err
+	}
+	if content != nil {
+		consumer, err := sourcemap.Parse(sourceMapURL, content)
+		if err != nil {
+			store.cache[cacheKey] = nil
+			level.Debug(store.l).Log("msg", "failed to parse sourcemap", "url", sourceMapURL, "release", release, "err", err)
+			return nil, err
+		}
+		level.Info(store.l).Log("msg", "successfully parsed sourcemap", "url", sourceMapURL, "release", release)
+		smap := &sourceMap{
+			consumer: consumer,
+		}
+		store.cache[cacheKey] = smap
+		store.metrics.cacheSize.WithLabelValues(getOrigin(sourceURL)).Inc()
+		return smap, nil
+	}
+	return nil, nil
+}
+
+// ResolveSourceLocation resolves minified source location to original source location
+func (store *RealSourceMapStore) ResolveSourceLocation(frame *models.Frame, release string) (*models.Frame, error) {
+	smap, err := store.getSourceMap(frame.Filename, release)
+	if err != nil {
+		return nil, err
+	}
+	if smap == nil {
+		return nil, nil
+	}
+
+	file, function, line, col, ok := smap.consumer.Source(frame.Lineno, frame.Colno)
+	if !ok {
+		return nil, nil
+	}
+	// unfortunately in many cases go-sourcemap fails to determine the original function name.
+	// not a big issue as long as file, line and column are correct
+	if len(function) == 0 {
+		function = "?"
+	}
+	return &models.Frame{
+		Filename: file,
+		Lineno:   line,
+		Colno:    col,
+		Function: function,
+	}, nil
+}
+
+// TransformException will attempt to resolved all monified source locations in the stacktrace with original source locations
+func (store *RealSourceMapStore) TransformException(ex *models.Exception, release string) *models.Exception {
+	if ex.Stacktrace == nil {
+		return ex
+	}
+	frames := []models.Frame{}
+
+	for _, frame := range ex.Stacktrace.Frames {
+		mappedFrame, err := store.ResolveSourceLocation(&frame, release)
+		if err != nil {
+			level.Error(store.l).Log("msg", "Error resolving stack trace frame source location", "err", err)
+			frames = append(frames, frame)
+		} else {
+			if mappedFrame != nil {
+				frames = append(frames, *mappedFrame)
+			} else {
+				frames = append(frames, frame)
+			}
+		}
+	}
+
+	return &models.Exception{
+		Type:       ex.Type,
+		Value:      ex.Value,
+		Stacktrace: &models.Stacktrace{Frames: frames},
+		Timestamp:  ex.Timestamp,
+	}
+}
+
+func cleanForFilePath(x string) string {
+	return strings.TrimLeft(strings.ReplaceAll(strings.ReplaceAll(x, "\\", ""), "/", ""), ".")
```

Comment:
The x string needs to be [cleaned](https://pkg.go.dev/path#Clean) to prevent [path traversal](https://owasp.org/www-community/attacks/Path_Traversal). In this case it won't pose an actual problem as the path undergoes a `path.Join` function afterwards which will cleanup the path, but I would add it as an extra security measure.

Reply:
This is for sanitizing a single path item rather than the entire path, actual sanitization for path traversal is happening [here](https://github.com/grafana/agent/blob/563dad3763ccad9823b7747dbef8a71427901dcc/pkg/integrations/v2/app_o11y_receiver/sourcemaps/sourcemaps.go#L195).


`path.Clean` **does not prevent path traversal**, they really should put a sticker warning on it :/ See http://technosophos.com/2016/03/31/go-quickly-cleaning-filepaths.html.

Reply:
- [ ] improve function name to mention path part

Reply:
thanks for  looking out for this 👍  great review